### ISSUE TITLE
Katib job and Katib services cannot be in different namespaces due to metrics collector service account

### ISSUE THREAD

**Issue by xyhuang** [kubeflow/kubeflow#2311]
every katib job comes with its own metrics collector, which is supposed to be in the same name space as the job. However the metrics collectors rely on a service account that is created at the time of katib service installation, which is bound to the namespace of katib services. This prevents katib job from running in a namespace that's different from katib services.

**Comment by xyhuang** [comment]
- do we assume that Katib is namespace scoped and jobs only work in same namespace
- or shall we create a separate service account for each metrics collector/namespace?
- any other possible solutions?
WDYT @YujiOshima @johnugeorge

**Comment by janvdvegt** [comment]
From a technical perspective I cannot contribute but for our use case splitting these up would be very benificial

**Comment by jlewi** [comment]
Do we have a design doc that covers how metrics will be collected from Jobs?

**Comment by johnugeorge** [comment]
@YujiOshima Do you have some version of Katib design doc? I will help you to make it up-to-date

**Comment by hougangliu** [comment]
@xyhuang when you try to create a studyjob in non-kubeflow namespace, you need run below command. I will update user doc for this guideline
```
$ cat mcrbac.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: metrics-collector
  namespace: nonkubeflow
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: metrics-collector
  namespace: nonkubeflow
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: metrics-collector
subjects:
- kind: ServiceAccount
  name: metrics-collector
  namespace: nonkubeflow
$ kubeflow apply -f  mcrbac.yaml
```

**Comment by xyhuang** [comment]
thanks @hougangliu! this is a valid workaround, though i wonder if there could be a "cleaner" way to fix that :) because running a separate RBAC creation file might potentially make it inconsistent with the codes [here](https://github.com/kubeflow/kubeflow/blob/c416045fb4b1458d5cd6bb4f28146d31a476a793/kubeflow/katib/studyjobcontroller.libsonnet#L29) if future changers are not careful enough to keep both places synced, in that case we might end up using different RBAC rules in different namespace.

**Comment by hougangliu** [comment]
Because we cannot know in advance which namespace user will run a studyjob, so I am afraid we cannot create a complete RBAC when install.
As you mentioned, this is indeed a problem to keep sync. Maybe we can assign **create** permission of **serviceaccount** and **RoleBinding** to  serviceaccount  studyjob-controller when install katib. Then when create a studyjob, study-controller try to create resources like mcrbac.yaml above if in need.

**Comment by johnugeorge** [comment]
Yes.  it is difficult to support metrics collector service account creation(during katib install) in this architecture design.  I feel that this metric collector architecture is little complex.

**Comment by jlewi** [comment]
What is the remaining work to be done with this issue? Is there work we plan on completing in 0.5?

**Comment by johnugeorge** [comment]
@jlewi With the current metric collector design, it is difficult to solve.  We have to keep this open till we move into a push based model 
/cc @richardsliu

**Comment by richardsliu** [comment]
Since this involves redesigning the metric collector, it should be part of the v1alpha2 design.

**Comment by gaocegege** [comment]
@johnugeorge Will we implement the push-based model in v1alpha2?

**Comment by hougangliu** [comment]
> @johnugeorge Will we implement the push-based model in v1alpha2?

After integration with [metadata](https://github.com/kubeflow/metadata) in v1alpha2, push-based model seems not in need any more

**Comment by gaocegege** [comment]
@hougangliu Can you please explain more about it? I think we still need to get and push metrics although we have metadata as a unified storage backend.

**Comment by johnugeorge** [comment]
@gaocegege No. Our earlier plan was to have a push based model when metadata project comes with python api to post metrics. 

@hougangliu Which integration are you talking about?

**Comment by hougangliu** [comment]
integration with metadata in v1alpha2 will not be in release 0.6, but I think in long term, we must implement the integration, replacing current metricCollector

**Comment by gaocegege** [comment]
I prefer to provide an interface similar to CRI, maybe the name could be MetricsCollectionAPI for early stopping services. Then we could support different backends. Because there are some users maybe not use metadata. We could keep extensibility for the external storage.

**Comment by gaocegege** [comment]
We can discuss it after v1alpha2 0.6

**Comment by jlewi** [comment]
@johnugeorge @gaocegege @richardsliu Do we need this issue to be fixed for 0.6?

**Comment by richardsliu** [comment]
We didn't have a nice way to fix this because we are still using the metrics collector implementation from v1alpha1. We decided to punt this until we can find a different solution for metrics collection (e.g. metadata APIs).

**Comment by jlewi** [comment]
@xyhuang @johnugeorge @richardsliu @gaocegege Any plans to fix this is part of 0.7? Do we need to redesign Katib to have better multi-user support?

Starting with 0.6 Katib shared services should be in the kubeflow namespace but users should launch their jobs in their own kubeflow-${USER} namespace? Do we need to make significant changes to Katib in order to support this?

**Comment by hougangliu** [comment]
@jlewi yes, 0.7 will resolve this problem (https://github.com/kubeflow/katib/issues/685)
